zerver/lib/actions.py

```diff
@@ -4656,9 +4644,15 @@ def do_update_message(user_profile: UserProfile, message: Message,
         if new_stream is not None:
             assert stream_being_edited is not None
 
-            if subs_gaining_usermessages:
+            if not new_stream.is_history_public_to_subscribers():
+                # For private streams, with history not public to subscribers,
+                # we find out users who didn't receive the message (by checking UserMessage rows)
+                # and create new UserMessage for these users so that they can
+                # access this message.
                 ums_to_create = []
                 for msg in changed_messages:
+                    user_ids_for_which_ums_exist = UserMessage.objects.select_related().filter(message__id=msg.id).values_list('user_profile__id', flat=True)
```

Comment:
@timabbott would this be too expensive? We need this successfully implement this feature.

Reply:
Changed this into a bulk query. Should be fine.